src/Oscoin/Crypto/Hash/RealWorld.hs

```diff
@@ -0,0 +1,125 @@
+
+{------------------------------------------------------------------------------
+  Crypto-specific instances
+------------------------------------------------------------------------------}
+
+instance HasHashing RealWorldCrypto where
+    type HashAlgorithm RealWorldCrypto = Blake2b_256
+
+    -- | A 'Digest' obtained by some hash algorithm @(HashAlgorithm RealWorldCrypto)@.
+    --
+    -- Textual representations use base-58 encoding. Serialisation and
+    -- deserialisation via "Crypto.Hash.Multi".
+    newtype Hash RealWorldCrypto =
+        Hash { fromHash :: Digest (HashAlgorithm RealWorldCrypto) }
+        deriving (Eq, Ord, ByteArrayAccess)
+
+    fromByteArray = Hash . Crypto.hash
+    toByteArray = convert . fromHash
+
+    hashAlgorithm = Blake2b_256
+
+    zeroHash = Hash . fromJust $
+        Crypto.digestFromByteString @(HashAlgorithm RealWorldCrypto) @ByteString
+            (ByteArray.zero (hashDigestSize (Proxy @(HashAlgorithm RealWorldCrypto))))
+
+{------------------------------------------------------------------------------
+  Instances galore
+-------------------------------------------------------------------------------}
+
+instance Hashable RealWorldCrypto Text where
+    hash = toHashed . Hash . Crypto.hash . encodeUtf8
+
+instance Hashable RealWorldCrypto ByteString where
+    hash = toHashed . Hash . Crypto.hash
+
+instance Hashable RealWorldCrypto LByteString where
+    hash = toHashed . fromHashed . hash . LBS.toStrict
+
+instance Hashable RealWorldCrypto Word8 where
+    hash = toHashed . Hash . Crypto.hash . BS.singleton
+
+instance (ByteArrayAccess a) => Hashable RealWorldCrypto (Maybe a) where
+    hash (Just x) = toHashed . Hash $ Crypto.hash x
+    hash Nothing  = toHashed $ zeroHash
+
+instance Multihashable (HashAlgorithm RealWorldCrypto) => FromJSON (Hash RealWorldCrypto) where
+    parseJSON = withText "Hash" $
+        either fail pure
+            . second Hash . decodeAtBase BaseN.Base58 . encodeUtf8
+
+instance Sql.FromField (Hash RealWorldCrypto) where
+    fromField f =
+        case Sql.fieldData f of
+            Sql.SQLText t ->
+                either (const sqlErr) (Sql.Ok . Hash)
+                       (decodeAtBase BaseN.Base58 $ encodeUtf8 t)
+            _ ->
+                sqlErr
+      where
+        sqlErr = Sql.returnError Sql.ConversionFailed f
+            "couldn't convert from Base58"
+
+instance Multihashable (HashAlgorithm RealWorldCrypto) => Serialise (Hash RealWorldCrypto) where
+    encode = Multihash.CBOR.encode . fromHash
+    decode = Hash <$> Multihash.CBOR.decode
+
+instance Multihashable (HashAlgorithm RealWorldCrypto) => FromHttpApiData (Hash RealWorldCrypto) where
+    parseQueryParam =
+        bimap T.pack Hash . decodeAtBase BaseN.Base58 . encodeUtf8
+
+
+instance Buildable (Hash RealWorldCrypto) where
+    build (Hash digest) =
+        let (bs :: ByteString) = convert digest
+        in BaseN.encodedTextBuilder $ BaseN.encodeAtBase BaseN.Base58 bs
```

Comment:
I think this now introduced subtle bugs: `parseQueryParam . toQueryParam /= id`, `fromField . toField /= id`, `parseJSON . toJSON /= id`. 

Shall we just define `Buildable (Hash RealWorldCrypto)` to give us a Base58-encoded _multihash_, and define all other stringly instances in terms of `build`?

Reply:
I am quite toasted after a day spent in the trenches, but what you propose sounds plausible. What needs to be worked out is whether or not the `Multihash` constraint ends up being "infectious" and starts to bubble up in functions which now requires _only_ a `Buildable (Hash c)` instance. That would be bad because we do not have an `instance Multihash (Hash MockCrypto)` and we cannot even define it (for the reasons we discussed previously).

I need to take a fresh look tomorrow morning.

Reply:
The `Mutihashable` is determined by `Hash RealWorldCrypto`, so methinks the `instance Buildable (Hash MockCrypto)` could just do something else, provided the `From*`-type instances are on the concrete types.